app/services/katello/applicability/applicable_content_helper.rb

```diff
@@ -92,6 +96,20 @@ def fetch_rpm_content_ids
             :enabled_module_streams => enabled_module_stream_ids).pluck(:id).uniq
       end
 
+      def fetch_newest_distinct_installed_packages
+        newest_distinct_installed_packages_query =
+          "SELECT DISTINCT ON (name) katello_installed_packages.* " \
+          "FROM katello_installed_packages INNER JOIN " \
+          "katello_host_installed_packages ON " \
+          "katello_installed_packages.id = " \
+          "katello_host_installed_packages.installed_package_id " \
+          "WHERE katello_host_installed_packages.host_id = " \
+          ":content_facet_id ORDER BY name, evr DESC"
+
+        ::Katello::InstalledPackage.find_by_sql([newest_distinct_installed_packages_query,
+                                                 content_facet_id: content_facet.host.id]).map(&:id)
```

Comment:
could you pass the sql straight  into the query above?  
```             katello_installed_packages.id in (#{newest_distinct_installed_packages.join(",")})").```

instead of evaluating it here and then passing the results?

Reply:
actually, how does this query work exactly?  I don't see it doing a > or any comparison to get the latest (i'm probably overlooking it)

Reply:
Sure, I'll give it a go.

Reply:
Hm, I'm having an issue `ERROR:  subquery has too many columns`. Perhaps this query is growing too large?

I tried this:

```ruby
::Katello::Rpm.
          joins("INNER JOIN katello_repository_rpms ON
            katello_rpms.id = katello_repository_rpms.rpm_id").
          joins("INNER JOIN katello_installed_packages ON
            katello_rpms.name = katello_installed_packages.name AND
            katello_rpms.arch = katello_installed_packages.arch AND
            katello_rpms.evr > katello_installed_packages.evr").
            #katello_installed_packages.id in (#{fetch_newest_distinct_installed_packages})").
          joins("LEFT JOIN katello_module_stream_rpms ON
            katello_rpms.id = katello_module_stream_rpms.rpm_id").
          joins("INNER JOIN katello_host_installed_packages ON
            katello_installed_packages.id = katello_host_installed_packages.installed_package_id").
          where("katello_repository_rpms.repository_id in (:bound_library_repos)",
            :bound_library_repos => self.bound_library_instance_repos).
          where("katello_host_installed_packages.host_id = :content_facet_id",
            :content_facet_id => self.content_facet.host.id).
          where("katello_installed_packages.id in (SELECT DISTINCT ON (katello_installed_packages.name) katello_installed_packages.id, katello_installed_packages.name, katello_installed_packages.evr FROM katello_installed_packages ORDER BY katello_installed_packages.name, katello_installed_packages.evr DESC)").
          where("katello_module_stream_rpms.module_stream_id is null or
            katello_module_stream_rpms.module_stream_id in (:enabled_module_streams)",
            :enabled_module_streams => enabled_module_stream_ids).pluck(:id).uniq
```

Reply:
I think that inner select has to only return an id, i think something like this works? 

select distinct on (name) id  from blahblahblha

Reply:
Okay, I've addressed this.